services/publishing/Tweek.Publishing.Service/Helpers/HooksHelper.cs

```diff
@@ -0,0 +1,153 @@
+
+        var keyPathsByHook = _AggregateKeyPathsByHook(keyPaths, allHooks);
+        var usedKeyPaths = _GetUsedKeyPaths(keyPathsByHook);
+        var keyPathsData = await _GetKeyPathsData(usedKeyPaths, commitId);
+
+        var hooksWithData = _GetHooksWithKeyPathData(keyPathsByHook, keyPathsData);
+        await _TriggerHooks(hooksWithData, commitId);
+      } catch (Exception ex) {
+        _logger.LogError(ex, $"Failed triggering notification hooks for commit {commitId}");
+        _metrics.Measure.Counter.Increment(_hooksMetric, _metricsFailure);
+      }
+    }
+
+    private async Task _TriggerHooks(Dictionary<Hook, string> hooksWithData, string commitId) {
+      var triggerTasks = hooksWithData.Select( kvp => kvp.Key.Trigger(kvp.Value) );
+
+      foreach (var triggerTask in triggerTasks) {
+        try {
+          await triggerTask;
+
+          _metrics.Measure.Counter.Increment(_hooksMetric, _metricsSuccess);
+        } catch (Exception ex) {
+          _logger.LogError(ex, $"Failed triggering a notification hook for commit {commitId}");
+          _metrics.Measure.Counter.Increment(_hooksMetric, _metricsFailure);
+        }
+      }
+    }
+
+    private Dictionary<Hook, string> _GetHooksWithKeyPathData(
+      Dictionary< Hook, HashSet<string> > keyPathsByHook,
+      Dictionary<string, KeyPathData> keyPathsData
+    ) {
+      return keyPathsByHook.Aggregate(new Dictionary<Hook, string>(), (acc, kvp) => {
+        var hookData = kvp.Value.Select( keyPath => keyPathsData[keyPath] );
+        var hookDataJson = JsonConvert.SerializeObject(hookData);
+
+        acc.Add(kvp.Key, hookDataJson);
+        return acc;
+      });
+    }
+
+    private async Task< Dictionary<string, KeyPathData> > _GetKeyPathsData(IEnumerable<string> keyPaths, string commitId) {
+      var keyPathsDataDict = new Dictionary<string, KeyPathData>(keyPaths.Count());
+
+      var keyDataTasks = keyPaths.Select(keyPath => new Task<string>[] {
+        Task.FromResult(keyPath),
+        _git($"show {commitId}:implementations/jpad/{keyPath}.jpad"),
+        _git($"show {commitId}:manifests/{keyPath}.json")
+      });
+
+      foreach (var tasks in keyDataTasks) {
+        var keyPathData = new KeyPathData(await tasks[0], await tasks[1], await tasks[2]);
+        keyPathsDataDict.Add(keyPathData.keyPath, keyPathData);
+      }
+
+      return keyPathsDataDict;
+    }
+
+    private IEnumerable<string> _GetUsedKeyPaths(Dictionary< Hook, HashSet<string> > keyPathsByHook) {
+      return keyPathsByHook.Values.Aggregate(new HashSet<string>(), ( keyPathsAcc, currentKeyPaths ) => {
+        keyPathsAcc.UnionWith(currentKeyPaths);
+        return keyPathsAcc;
+      });
+    }
+
+    private Dictionary< Hook, HashSet<string> > _AggregateKeyPathsByHook(IEnumerable<string> allKeyPaths, KeyHooks[] allHooks) {
+      var hooksDict = new Dictionary< Hook, HashSet<string> >();
+
+      return allHooks.Aggregate(hooksDict, (hooks, keyHooks) => {
```

Comment:
Can be simpler, maybe something like:
allHooks.Select(x=> (hook: x, paths: x.GetMatchingKeyPaths(allKeyPaths))).ToDictionary(x=> x.hook, x.paths)

Reply:
Its a bit more complicated than the suggestion for 3 reasons:

1. `allHooks` is actually an array of KeyHooks (each containing a keyPath and an array of Hooks for that keyPath). I've renamed all the `allHooks` variables in the code to a better name to make that clearer.
2. We need to remove hooks that don't match any keyPath from this commit from the results
3. A `Hook` instance is the combination of `type` and `url` to notify. The same hook might appear in several different KeyHooks (if they want different keys to notify the same url), and this function groups by the hook itself and not the KeyHook.
This means the `ToDictionary` will get duplicate keys (since `Hook` is the key) which it can't handle, so we need to first group the keyPaths by hook before we can convert it to a dictionary (and remove duplicate keyPaths from each hook if there are any).

Not sure if that explanation is understandable, so in case it wasn't, the example in the unit test for `AggregateKeyPathsByHook` might make it clearer.

Incorporating these points into your suggestion I came up with:

```C#
return allKeyHooks
  .SelectMany(keyHooks => {
    var keyPaths = keyHooks.GetMatchingKeyPaths(allKeyPaths);
    return keyHooks.GetHooks().Select( hook => ( hook, keyPaths ) );
  })
  .Where( hookTuple => hookTuple.keyPaths.Count() > 0 )
  .GroupBy( hookTuple => hookTuple.hook, hookTuple => hookTuple.keyPaths )
  .ToDictionary(
    hookGrouping => hookGrouping.Key,
    hookGrouping => hookGrouping.SelectMany(keyPaths => keyPaths).ToHashSet()
  );
```

It is slightly shorter than the previous `Aggregate` code, though it might be a bit more complicated than it. WDYT?

I'm including the previous `Aggregate` code here for reference:

```C#
var hooksDict = new Dictionary< Hook, HashSet<string> >();

return allHooks.Aggregate(hooksDict, (hooks, keyHooks) => {
  var keyPaths = keyHooks.GetMatchingKeyPaths(allKeyPaths);
  if (keyPaths.Count() == 0) return hooks;

  foreach (var hook in keyHooks.GetHooks()) {
    if (!hooks.ContainsKey(hook)) hooks.Add(hook, keyPaths.ToHashSet<string>());

    hooks[hook].UnionWith(keyPaths.ToHashSet<string>());
  }

  return hooks;
});
```